src/cls/rgw/cls_rgw_types.h

```diff
@@ -1123,18 +1123,19 @@ struct cls_rgw_obj {
   cls_rgw_obj(std::string& _p, cls_rgw_obj_key& _k) : pool(_p), key(_k) {}
 
   void encode(ceph::buffer::list& bl) const {
-    ENCODE_START(2, 1, bl);
+    ENCODE_START(3, 1, bl);
     encode(pool, bl);
-    encode(key.name, bl);
```

Comment:
Not a very elegant way to do it - but what if we encode an empty string here, in place of removing key.name altogether? We might just leave the version to 2 and while decoding key.name will be first set to an empty string, but later will be set to the correct value?

Reply:
I was thinking of something like this:

void encode(ceph::buffer::list& bl) const {
    ENCODE_START(2, 1, bl);
    encode(pool, bl);
    encode("", bl);
    encode(loc, bl);
    encode(key, bl);
    ENCODE_FINISH(bl);
  }

  void decode(ceph::buffer::list::const_iterator& bl) {
    DECODE_START(2, bl);
    decode(pool, bl);
    decode(key.name, bl);
    decode(loc, bl);
    if (struct_v >= 2) {
      decode(key, bl);
    }
    DECODE_FINISH(bl);
  }

Where we don't touch the version at all. Will this not be allowed?

Reply:
by bumping that second argument to `ENCODE_START()` to 2, we'd cause v1 decoders to fail with `decode no longer understand old encoding version 1 < 2`. and v1 is old enough by now that we can safely assume we don't need to support it - we don't promise backward compatibility over more than two major releases

i'm not sure that we need to bump the first argument from 2 to 3 here. but i'm trying to find a way to eventually remove this field entirely, and i *think* it could help with that

for example, maybe we plan ahead for a v4 that removes this field, but only bump to v3 for now:
```c++
  void decode(ceph::buffer::list::const_iterator& bl) {
    DECODE_START(3, bl);
    decode(pool, bl);
    if (struct_v < 4) { // plan to remove this in v4
      decode(key.name, bl);
    }
    decode(loc, bl);
    if (struct_v >= 2) {
      decode(key, bl);
    }
    DECODE_FINISH(bl);
  }
```
with that decoder support in place, we could follow up two release later with an updated encoder:
```c++
  void encode(ceph::buffer::list& bl) const {
    ENCODE_START(4, 3, bl);
    encode(pool, bl);
    encode(loc, bl);
    encode(key, bl);
    ENCODE_FINISH(bl);
  }
```
i think that could work.. but it ties our hands to this version 4, and prevents us from making other encoding changes here in the meantime

Reply:
if we backport the v3 changes to pacific/quincy, we might even consider merging v4 for reef

Reply:
> by bumping that second argument to `ENCODE_START()` to 2, we'd cause v1 decoders to fail with `decode no longer understand old encoding version 1 < 2`. and v1 is old enough by now that we can safely assume we don't need to support it - we don't promise backward compatibility over more than two major releases
> 
> i'm not sure that we need to bump the first argument from 2 to 3 here. but i'm trying to find a way to eventually remove this field entirely, and i _think_ it could help with that
> 
> for example, maybe we plan ahead for a v4 that removes this field, but only bump to v3 for now:
> 
> ```c++
>   void decode(ceph::buffer::list::const_iterator& bl) {
>     DECODE_START(3, bl);
>     decode(pool, bl);
>     if (struct_v < 4) { // plan to remove this in v4
>       decode(key.name, bl);
>     }
>     decode(loc, bl);
>     if (struct_v >= 2) {
>       decode(key, bl);
>     }
>     DECODE_FINISH(bl);
>   }
> ```
> 
> with that decoder support in place, we could follow up two release later with an updated encoder:
> 
> ```c++
>   void encode(ceph::buffer::list& bl) const {
>     ENCODE_START(4, 3, bl);
>     encode(pool, bl);
>     encode(loc, bl);
>     encode(key, bl);
>     ENCODE_FINISH(bl);
>   }
> ```
> 
> i think that could work.. but it ties our hands to this version 4, and prevents us from making other encoding changes here in the meantime

If we need to make further encoding changes, we might use version 5, and things will still work.

Reply:
> if we backport the v3 changes to pacific/quincy, we might even consider merging v4 for reef

we could do that provided the backports to pacific/quincy happen asap.

Reply:
> > if we backport the v3 changes to pacific/quincy, we might even consider merging v4 for reef
> 
> we could do that provided the backports to pacific/quincy happen asap.

No, this might become a problem, if we do not want to remove encoding of key.name in version 5.

Reply:
> > by bumping that second argument to `ENCODE_START()` to 2, we'd cause v1 decoders to fail with `decode no longer understand old encoding version 1 < 2`. and v1 is old enough by now that we can safely assume we don't need to support it - we don't promise backward compatibility over more than two major releases
> > i'm not sure that we need to bump the first argument from 2 to 3 here. but i'm trying to find a way to eventually remove this field entirely, and i _think_ it could help with that
> > for example, maybe we plan ahead for a v4 that removes this field, but only bump to v3 for now:
> > ```c++
> >   void decode(ceph::buffer::list::const_iterator& bl) {
> >     DECODE_START(3, bl);
> >     decode(pool, bl);
> >     if (struct_v < 4) { // plan to remove this in v4
> >       decode(key.name, bl);
> >     }
> >     decode(loc, bl);
> >     if (struct_v >= 2) {
> >       decode(key, bl);
> >     }
> >     DECODE_FINISH(bl);
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with that decoder support in place, we could follow up two release later with an updated encoder:
> > ```c++
> >   void encode(ceph::buffer::list& bl) const {
> >     ENCODE_START(4, 3, bl);
> >     encode(pool, bl);
> >     encode(loc, bl);
> >     encode(key, bl);
> >     ENCODE_FINISH(bl);
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i think that could work.. but it ties our hands to this version 4, and prevents us from making other encoding changes here in the meantime
> 
> If we need to make further encoding changes, we might use version 5, and things will still work.

No, this might become a problem, if we do not want to remove encoding of key.name in version 5.